### PR TITLE
Add Stellar support

### PR DESCRIPTION
**Pull request by madcake**
## Checklist

- [x] Add tests to cover changes as needed.
- [x] Return correct curve and purpose in `src/Coin.cpp`.
- [x] Derivation from private key in `src/Coin.cpp`.
- [x] Validation in `src/Coin.cpp`.

### CODE REVIEW

File: src/Stellar/Address.cpp [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    // Calculate checksum for existing bytes
+    uint16_t crc = 0x0000;
+    uint16_t polynomial = 0x1021;
+    uint32_t i;
+    uint8_t bit;
+    uint8_t byte;
+    uint8_t bitidx;
+    uint8_t c15;
+
+    for (i=0; i < length; i++) {
```
Comment (alejandro-isaza) [inline src/Stellar/Address.cpp]:
```suggestion
    for (auto i = 0; i < length; i++) {
```

File: src/Stellar/Address.cpp [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        byte = bytes[i];
```
Comment (alejandro-isaza) [inline src/Stellar/Address.cpp]:
```suggestion
        auto byte = bytes[i];
```

File: src/Stellar/Address.cpp [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        for (bitidx=0; bitidx < 8; bitidx++) {
```
Comment (alejandro-isaza) [inline src/Stellar/Address.cpp]:
```suggestion
        for (auto bitidx = 0; bitidx < 8; bitidx++) {
```

File: src/Stellar/Address.cpp [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+            bit = ((byte >> (7 - bitidx) & 1) == 1);
```
Comment (alejandro-isaza) [inline src/Stellar/Address.cpp]:
```suggestion
            auto bit = ((byte >> (7 - bitidx) & 1) == 1);
```

File: src/Stellar/Address.h [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+#include <string>
+
+namespace TW {
+namespace Stellar {
+
+class Address {
+public:
+
+    // 56 character base-32 encoded string
+    static const size_t STELLAR_ADDRESS_SIZE = 56;
```
Comment (alejandro-isaza) [inline src/Stellar/Address.h]:
```suggestion
    static const size_t size = 56;
```

File: src/Stellar/Address.h [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    // Decodes to 35 bytes
+    static const size_t STELLAR_ADDRESS_SIZE_RAW = 35;
```
Comment (alejandro-isaza) [inline src/Stellar/Address.h]:
```suggestion
    static const size_t rawSize = 35;
```

File: src/Stellar/Address.h [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    // Raw key size is 32 bytes
+    static const size_t STELLAR_KEY_SIZE = 32;
```
Comment (alejandro-isaza) [inline src/Stellar/Address.h]:
```suggestion
    static const size_t keySize = 32;
```

File: src/Stellar/Address.cpp [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+    bytes_full[keylen-2] = checksum & 0x00ff;
+    bytes_full[keylen-1] = (checksum>>8) & 0x00ff;
+
+    base32_encode(bytes_full, keylen, out, sizeof(out), BASE32_ALPHABET_RFC4648);
+
+    // Public key will always be 56 characters
+    return std::string(out);
+}
+
+uint16_t Address::crc16(uint8_t *bytes, uint32_t length) {
```
Comment (hewigovens) [inline src/Stellar/Address.cpp]: Let’s move crc16 out of this address implementation? IOST public key pair encoding requires crc32, so it’s better to do it like class Hash, put crc16 / crc32 into one place

File: src/Stellar/Address.cpp [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+    if (string.length() != STELLAR_ADDRESS_SIZE) {
+        return false;
+    }
+
+    // Check that it decodes correctly
+    uint8_t *ret = base32_decode(string.data(), STELLAR_ADDRESS_SIZE, decoded, sizeof(decoded), BASE32_ALPHABET_RFC4648);
+    valid = (ret != NULL);
+
+    // ... and that version byte is 0x30
+    if (valid && decoded[0] != 0x30) {
```
Comment (hewigovens) [inline src/Stellar/Address.cpp]: Use const variable or enum for 0x30?

File: trezor-crypto/CMakeLists.txt [hunk 10]
```diff
@@ -22,8 +22,7 @@ add_library(TrezorCrypto
     src/cash_addr.c
     src/memzero.c
     src/scrypt.c
-    src/ripple/base58.c
-)
+    src/ripple/base58.c)
```
Comment (hewigovens) [inline trezor-crypto/CMakeLists.txt]: Why change this format?

Comment (madcake) [reply trezor-crypto/CMakeLists.txt]: I was try out crc to class and cleaned extra line as on java :).

File: src/interface/TWCoinTypeConfiguration.cpp [hunk 11]
```diff
@@ -138,6 +142,7 @@ const char *explorerURLForCoinType(enum TWCoinType type) {
     case TWCoinTypeBinance: return "https://binance.com";
     case TWCoinTypeEOS: return "https://eospark.com";
     case TWCoinTypeRipple: return "https://bithomp.com";
+    case  TWCoinTypeStellar: return "https://stellarscan.io";
```
Comment (vikmeup) [inline src/interface/TWCoinTypeConfiguration.cpp]: @madcake add tests for this

**Comment by vikmeup** [comment]
@madcake can you pull the latest master and update on the status for address derivation?

File: include/TrustWalletCore/TWHDWallet.h [hunk 12]
```diff
@@ -49,6 +49,10 @@ TWData *_Nonnull TWHDWalletSeed(struct TWHDWallet *_Nonnull wallet);
 TW_EXPORT_PROPERTY
 TWString *_Nonnull TWHDWalletMnemonic(struct TWHDWallet *_Nonnull wallet);
 
+/// Generates the private key at the specified derivation path.
+TW_EXPORT_METHOD
+struct TWPrivateKey *_Nonnull TWHDWalletGetAccountKey(struct TWHDWallet *_Nonnull wallet, enum TWCoinType coin, uint32_t account);
```
Comment (hewigovens) [inline include/TrustWalletCore/TWHDWallet.h]: simply `TWHDWalletGetKey`?

Comment (madcake) [reply include/TrustWalletCore/TWHDWallet.h]: Cannot overload a function.

File: src/Stellar/Address.cpp (hunk 8 above)
Comment (hewigovens) [reply src/Stellar/Address.cpp]: Since we have `Crc.cpp`, this method can be removed

File: tests/interface/TWStellarTests.cpp [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+#include <TrustWalletCore/TWP2SHPrefix.h>
+#include <TrustWalletCore/TWPrivateKey.h>
+#include <TrustWalletCore/TWPublicKey.h>
+#include <TrustWalletCore/TWPurpose.h>
+#include <TrustWalletCore/TWStellarAddress.h>
+
+#include <gtest/gtest.h>
+#include <HDWallet.h>
+
+TEST(Stellar, DeriveTransparentAddress) {
```
Comment (hewigovens) [inline tests/interface/TWStellarTests.cpp]: Stellar has transparent things?

Comment (madcake) [reply tests/interface/TWStellarTests.cpp]: > Stellar has transparent things?

No it's copy paste :)

File: include/TrustWalletCore/TWStellarVersionByte.h [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+// terms governing use, modification, and redistribution, is contained in the
+// file LICENSE at the root of the source code distribution tree.
+
+#pragma once
+
+#include "TWBase.h"
+
+TW_EXTERN_C_BEGIN
+
+TW_EXPORT_ENUM(uint16_t)
```
Comment (alejandro-isaza) [inline include/TrustWalletCore/TWStellarVersionByte.h]: why it is `uint16_t` if it's called Version**Byte**?

File: src/Crc.cpp [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+
+using namespace TW;
+
+uint16_t Crc::crc16(uint8_t *bytes, uint32_t length) {
+    // Calculate checksum for existing bytes
+    uint16_t crc = 0x0000;
+    uint16_t polynomial = 0x1021;
+    uint32_t i;
+    uint8_t bit;
+    uint8_t byte;
```
Comment (alejandro-isaza) [inline src/Crc.cpp]: a few of these are unused. please remove.

**Pull request merged**